cmd/gosedctl/cmd.go

```diff
@@ -0,0 +1,123 @@
+
+	"golang.org/x/crypto/pbkdf2"
+)
+
+type context struct{}
+
+type initialSetupCmd struct {
+	Device   string `flag:"" required:"" short:"d"  help:"Path to SED device (e.g. /dev/nvme0)"`
+	Password string `flag:"" optional:"" short:"p"`
+	Hash     string `flag:"" optional:"" default:"sedutil-dta"`
+}
+
+var cli struct {
+	InitialSetup initialSetupCmd `cmd:"" help:"Take ownership of a given device"`
+}
+
+func (t *initialSetupCmd) Run(ctx *context) error {
+	fmt.Printf("Open device: %s", t.Device)
+	coreObj, err := core.NewCore(t.Device)
+	if err != nil {
+		return fmt.Errorf("NewCore(%s) failed: %v", t.Device, err)
+	}
+	fmt.Println("Find ComID")
+	comID, _, err := core.FindComID(coreObj.DriveIntf, coreObj.DiskInfo.Level0Discovery)
+	if err != nil {
+		return err
+	}
+	fmt.Println("Create new ControlSession")
+	cs, err := core.NewControlSession(coreObj.DriveIntf, coreObj.Level0Discovery, core.WithComID(comID))
+	if err != nil {
+		return fmt.Errorf("NewControllSession() failed: %v", err)
+	}
+
+	// Take Ownership
+	fmt.Println("Create new Session")
+	sess, err := cs.NewSession(uid.AdminSP)
+	if err != nil {
+		return fmt.Errorf("cs.NewSession() failed: %v", err)
+	}
+	defer sess.Close()
+
+	//Get the MSID (only works if device hasnt been claimed)
+	fmt.Println("Read MSID Pin")
+	msid, err := table.Admin_C_PIN_MSID_GetPIN(sess)
+	if err != nil {
+		return fmt.Errorf("Admin_C_PIN_MSID_GetPin() failed: %v", err)
+	}
+	// According to TCG_Storage_Opal_SSC_Application_Note_1-00_1-00-Final.pdf, p. 10 we have to close the session
+	// but this is not implemented. We use ThisSp_Authenticate to elevate the session directly.
+	fmt.Println("Authenticate with MSID as SID Authority at AdminSP")
+	if err := table.ThisSP_Authenticate(sess, uid.AuthoritySID, msid); err != nil {
+		return fmt.Errorf("ThisSp_Authenticate failed: %v", err)
+	}
+	fmt.Println("Set new password")
+	// Set the new SID password. Password needs to be hashed.
+	// The used algorithm is the same as used in DriveTrustAlliance implementation of sedutil-cli
+	serial, err := coreObj.SerialNumber()
+	if err != nil {
+		return fmt.Errorf("coreObj.SerialNumber() failed: %v", err)
+	}
+	salt := fmt.Sprintf("%-20s", serial)
+	pwhash := pbkdf2.Key([]byte(t.Password), []byte(salt[:20]), 75000, 32, sha1.New)
+
+	if err := table.Admin_C_Pin_SID_SetPIN(sess, pwhash); err != nil {
+		return fmt.Errorf("Admin_C_PIN_SID_SetPIN() failed: %v", err)
+	}
+
+	fmt.Println("Activate LockingSP")
+	// Activate LockingSP
+	lcs, err := table.Admin_SP_GetLifeCycleState(sess, uid.LockingSP)
+	if err != nil {
+		return fmt.Errorf("Admin_SP_GetLifeCycleState() failed: %v", err)
+	}
+	if lcs != table.ManufacturedInactive {
+		return fmt.Errorf("LockingSP Lifecycle state of %s, but require %s", lcs.String(), table.ManufacturedInactive)
+	}
+	if err := table.LockingSPActivate(sess); err != nil {
+		return fmt.Errorf("LockingSPActivate() failed: %v", err)
+	}
+	sess.Close()
```

Comment:
Will be closed by defer already

Reply:
Maybe split into separate functions from here on?

Reply:
I renamed it to adminSession and lockingSession to separate between the sessions and they are closed at the exact point they need to be closed.